File: frontend/src/api/smb.ts
import { client } from './client'

export interface SMBServer {
  id: number
  name: string
  host: string
  port: number
  share: string
  domain: string | null
  username: string
  created_at: string
}

export interface SMBEntry {
  name: string
  is_dir: boolean
  size: number
}

export interface CreateSMBPayload {
  name: string
  host: string
  port?: number
  share: string
  domain?: string
  username: string
  password: string
}

export const smbApi = {
  list: () => client.get<SMBServer[]>('/smb').then(r => r.data),
  create: (data: CreateSMBPayload) => client.post<SMBServer>('/smb', data).then(r => r.data),
  update: (id: number, data: Partial<CreateSMBPayload>) =>
    client.put<SMBServer>(`/smb/${id}`, data).then(r => r.data),
  delete: (id: number) => client.delete(`/smb/${id}`),
  test: (id: number) => client.post<{ ok: boolean; error?: string }>(`/smb/${id}/test`).then(r => r.data),
  browse: (id: number, path: string) =>
    client.get<SMBEntry[]>(`/smb/${id}/browse`, { params: { path } }).then(r => r.data),
}

File: backend/api/smb.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from api.deps import get_db
from core.crypto import encrypt
from models.smb_server import SMBServer
from schemas.smb_server import SMBServerCreate, SMBServerResponse, SMBServerUpdate
from smb.client import SMBClient, SMBConnectionError

router = APIRouter(prefix="/api/smb", tags=["smb"])


@router.get("", response_model=list[SMBServerResponse])
def list_servers(db: Session = Depends(get_db)):
    return db.query(SMBServer).all()


@router.post("", response_model=SMBServerResponse, status_code=201)
def create_server(data: SMBServerCreate, db: Session = Depends(get_db)):
    server = SMBServer(
        **data.model_dump(exclude={"password"}),
        password=encrypt(data.password),
    )
    db.add(server)
    db.commit()
    db.refresh(server)
    return server


@router.put("/{server_id}", response_model=SMBServerResponse)
def update_server(
    server_id: int,
    data: SMBServerUpdate,
    db: Session = Depends(get_db),
):
    server = db.query(SMBServer).filter(SMBServer.id == server_id).first()
    if not server:
        raise HTTPException(status_code=404, detail="Server not found")

    update_data = data.model_dump(exclude_none=True)
    if "password" in update_data:
        update_data["password"] = encrypt(update_data["password"])

    for key, value in update_data.items():
        setattr(server, key, value)
    db.commit()
    db.refresh(server)
    return server


@router.delete("/{server_id}", status_code=204)
def delete_server(server_id: int, db: Session = Depends(get_db)):
    server = db.query(SMBServer).filter(SMBServer.id == server_id).first()
    if not server:
        raise HTTPException(status_code=404, detail="Server not found")
    db.delete(server)
    db.commit()


@router.post("/{server_id}/test")
def test_connection(server_id: int, db: Session = Depends(get_db)):
    server = db.query(SMBServer).filter(SMBServer.id == server_id).first()
    if not server:
        raise HTTPException(status_code=404, detail="Server not found")

    try:
        client = SMBClient.from_server_model(server)
        client.connect()
        return {"ok": True}
    except SMBConnectionError as exc:
        return {"ok": False, "error": str(exc)}


@router.get("/{server_id}/browse")
def browse_directory(server_id: int, path: str = "/", db: Session = Depends(get_db)):
    server = db.query(SMBServer).filter(SMBServer.id == server_id).first()
    if not server:
        raise HTTPException(status_code=404, detail="Server not found")

    try:
        client = SMBClient.from_server_model(server)
        return client.list_directory(path)
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
